### DISCUSSION TITLE
Global Menu Widget

### DISCUSSION THREAD

**Discussion by skela** [qtile/qtile#3498] [Ideas]
Having a global menu widget would be fantastic.
In case its not known what that means, its the menu you typically see at the top  of the window decoration, you know, "File , Edit, View, etc" only that its moved away from the window, and the active window or window with focus has the menu displayed elsewhere (Mac OS is a pretty good example of global menus).

Some applications hide this menu and only show it when you press the Alt key (for example Firefox), but some windows, quite often electron based apps, dont do that, and the menu looks very strange in Qtile, because of the lack of window decorations.

Fildem, a global menu plugin for Gnome is written in python, so might be worth looking at that for inspiration.
https://github.com/gonzaarcr/Fildem

**Comment by m-col** [comment]
This would be pretty cool!

I think because of the use of dropdown menus and absence of a GUI widget toolkit that would support it, it's a bit out of scope for qtile proper. However I believe @elParaguayo has just that thing implemented already so it may well be a pretty simple task using his existing library for that at qtile-examples. Alternatively using Fildem's own code might well be feasible -- it looks like it has 2 parts, the gnome extension (im guessing this goes in the bar), and the GTK based dropdown menu component. Maybe only the first part would need to be re-implemented as a qtile widget, which then calls the second part.

The other thing that it would need would be hooking the window and menu UI management with something to listen on the interface - I don't see anything in the readme of Fildem about what that entails. If it's a dbus interface then it might be relatively simple, but if it's a gnome-specific interface then I don't think it would be possible outside of the gnome environment unfortunately.

**Reply by elParaguayo** [reply]
I avoided using the GTK library for building menus as I wanted a native implementation which could allow users to theme the menus in a similar way to how they configured widgets.

I would have thought the same library could be used here.

As for the menu, it feels like this would be very similar to the `StatusNotifier` widget: we register a service on dbus, apps then register themselves with this server and we can get the menu for the current window from the bus.

I like it as an intellectual exercise but I'm not sure I'd ever use it.

**Reply by elParaguayo** [reply]
Update: so, I've made a quick service and I can get apps to register their windows. It looks like the apps then present a dbusmenu object which *should* be compatible with the menus in `StatusNotifier`. If that's the case, then this may be pretty quick to knock out.

EDIT: I guess not every app is implementing this spec so it could be a bit hit and miss. The first window I looked at (`brave`) registers and provides details of its object on the bus but, when you use d-feet to check, it's not there... Next up, `code`, seems to be ok. `d-feet` and `alacritty` don't register at all.

**Reply by skela** [reply]
This is pretty awesome.

Yeah I reckon no matter what you do, you wont be able to do this for 100% of apps, but a 60%, 70% would be a pretty big improvement imho. 

if u can do it for code, maybe that means a few electron based apps will be sorted too, which is pretty epic if you ask me, because those are often the ones that don't do anything about it, and just leave that ugly bar visible at the top (code being the exception because at least they give u the option of hiding the bar).

Where's yer library @elParaguayo , would love to take a closer look or try it out.
Been dreaming about Global Menus in Qtile for a while now :P But just could not figure out how to use the fildem code :(

**Reply by elParaguayo** [reply]
https://github.com/elParaguayo/qtile-extras

**Reply by elParaguayo** [reply]
@skela - does fildem work with `code`? When I try to inspect the dbus interface with d-feet, I get the top level menu items (File, Edit etc) but I can't get anything in those actual menus.

I can see some comments in the code about having issues with electron apps but it's this sort of thing that really puts me off trying to implement something like this.

Ah... https://github.com/electron/electron/issues/8297

**Reply by skela** [reply]
@elParaguayo the fact that you managed to get the top level items is really cool, i didnt get that far :'( 
(this is where im at so far...)
![2022-04-12_21-48](https://user-images.githubusercontent.com/58268/163042160-a4b35e26-dda6-48ba-984c-8267b81fed4b.png)

**Reply by elParaguayo** [reply]
I've got top level menus and I know how to get the rest. It's just a case of piecing it together now...

**Reply by elParaguayo** [reply]
Just noticed your `TODO` comments.

Neither replacing the bar with the menu or keyboard navigation are possible (yet). This is a widget that takes up space in the bar.

**Reply by elParaguayo** [reply]
VLC's menu:
![2022-04-12-215258_589x35_scrot](https://user-images.githubusercontent.com/946265/163052757-9f5a09dc-d83a-4811-a9b9-a407ad09ec1b.png)

**Reply by elParaguayo** [reply]
Existing menu code looks like it works:
![2022-04-12-221129_1920x1080_scrot](https://user-images.githubusercontent.com/946265/163055394-97e2694a-7764-4138-ba1d-e7862e74a70a.png)

**Reply by skela** [reply]
@elParaguayo that looks awesome.

Yeah my TODO Was just a thought i had, how I could imagine doing it, I feel like I had a launcher at some point that replaced the bar, but then again that might be in xmonad. 

I'd be happy with having the menu show up at the top the way you have there too ❤️ As long as its removed from the window, ill be happy :D

**Reply by skela** [reply]
Your screenshot looks amazing! :D 

Do you know if apps that normally have the menu bar fixed to the top of the window, do those get moved out and moved down to the bar with your approach?

**Reply by elParaguayo** [reply]
It looks like, if the app registers with the widget, it then hides the menu bar from its own window.

**Reply by skela** [reply]
guess thats how it would have to work, and how kde plasma, fildem, and others, why it works for them.
thats awesome

**Reply by skela** [reply]
I don't know if this is helpful, but here's at least my list of apps that I can't wait to get the menu out from the window and into the bar:

- Inkscape
- GIMP
- OpenSCAD
- Ultimaker Cura (this one is probably never going to follow any kind of sensible standard)

**Reply by elParaguayo** [reply]
From what I can see, neither Inkscape nor GIMP implement this standard so the widget does nothing for those windows.

I don't have the other 2 installed.

**Reply by skela** [reply]
Hopefully you push yer menu experiments at some point :D cant wait to check it out :D

**Reply by elParaguayo** [reply]
Getting close to pushing something that you can try - it will have (many) bugs.

Just having some issues with focus at the moment - e.g. opening `code` registers menus for two different windows but it's not loading the menu until I refocus on the window which is just annoying.

**Reply by elParaguayo** [reply]
Fixed the loading problem. Next (odd) issue for `code` is that the menu doesn't respond to clicked items. Works beautifully with `vlc`...

**Reply by elParaguayo** [reply]
If you want to test it, you can try installing the `global-menu` branch from my repo: https://github.com/elParaguayo/qtile-extras/tree/global-menu

**Reply by skela** [reply]
Excellent :D Will give it a go :D

**Reply by skela** [reply]
Really amazing that you managed to do this in such a short span of time.
Works really well already :D It even managed to work for OpenSCAD, which i did not expect at all.

Worked flawlessly with the following apps that i tested:
OpenSCAD
Slack
VSCode
1Password (although theyve done something weird there so its not 100%).
Discord

Did not work in these apps:
GIMP (but u mentioned they do something weird so)
Inkscape (but u mentioned they do something weird so)
Thunar
Firefox
Thunderbird


https://user-images.githubusercontent.com/58268/163444265-427c02db-8364-482b-8ea1-6d4352623bc1.mp4

**Reply by elParaguayo** [reply]
Nice!

It helped that I had the menu code written already (because `StatusNotifierItems` use the same DBus menu interface).

For those apps that don't work, do you know whether they work with Fildem?

**Reply by skela** [reply]
I've not managed to get fildem to work via gnome extensions because the extension is for gnome shell 41 and ive got 42 installed so.
I'll let you know if I manage to get it running though.

**Reply by skela** [reply]
@elParaguayo have been using the global menu for a week now, absolutely loving it.
I like tinkering with things, so that I could possibly help via a fork, but just wondering if I've got something wrong in my setup, but does restarting qtile reload your global menu widget without any issues?

At the moment I have to logout and login, to get the menu to reload if i make a change, because restarting qtile, the menu disappears and does not want to come back.

**Reply by elParaguayo** [reply]
Forking would be great - very happy for you to have a play and suggest changes. I'd suggest you post questions on the `qtile-extras` github page rather than this one.

As for your second question, let me take a look.

**Reply by skela** [reply]
@elParaguayo I will going forward, just wasnt sure if u wanted me to post questions etc on the pull request or to create issues in that repo :D

**Reply by elParaguayo** [reply]
It's not merged yet so probably best to post comments, suggestions and issues in that PR thread.
